### PR TITLE
Ensure immutability of non-thunk values

### PR DESCRIPTION
**Pull request by edolstra**
# Motivation

Thunks are now overwritten by a helper function `Value::finishValue(newType, payload)` (where `payload` is the original anonymous union inside `Value`). This helps to ensure we never update a value elsewhere, since that would be incompatible with parallel evaluation (i.e. after a value has transitioned from being a thunk to being a non-thunk, it should be immutable).
    
There were two places where this happened: `Value::mkString()` and `ExprAttrs::eval()`.
    
This PR also adds a bunch of accessor functions for value contents, like `Value::integer()` to access the integer field in the union.
    
Bindings are now marked as "const" to prevent accidental modification.

# Context

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libexpr/value.hh [hunk 1]
```diff
@@ -286,34 +288,25 @@ public:
             abort();
     }
 
-    /**
-     * After overwriting an app node, be sure to clear pointers in the
-     * Value to ensure that the target isn't kept alive unnecessarily.
-     */
-    inline void clearValue()
+    inline void finishValue(InternalType newType, Payload newPayload)
```
Comment (roberth) [inline src/libexpr/value.hh:291]: Should it be like the following, or does the inlining take care of unnecessary copying? I would think that `newPayload` must still be the equivalent of a normal local variable after inlining.

```suggestion
    inline void finishValue(InternalType newType, Payload && newPayload)
```

**Review by roberth — COMMENTED** [review]
That was a big diff, but in general it looks good.

I appreciate your recent attention to safety, and we should definitely have some of that, but it feels a bit misaligned, considering that the evaluation process itself doesn't seem to have all that many issues at this time.
Don't get me wrong, I'm happy with this change, but I'd be even happier if it fixes a known bug.

**Comment by edolstra** [comment]
@roberth It's a big diff but the vast majority of it is fairly mechanical (e.g. changing `integer` to `integer()`).

The safety issue is a real one for multi-core evaluation, where functions like `Value::mkString()` modifying strings after setting their type really does lead to random crashes or different evaluation results.

**Comment by roberth** [comment]
> multi-core evaluation

That's new. Should we really bring that into scope?

**Review by tomberek — APPROVED** [review]

**Comment by roberth** [comment]
Needs rebase.

**Pull request merged**